core/src/main/scala/shapeless/constraints.scala

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2011-20 Miles Sabin 
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package shapeless
+
+import ops.hlist.{Selector => HSelector}
+import ops.coproduct.{Selector => CSelector}
+
+import scala.annotation.implicitNotFound
+
+/**
+ * Type class witnessing that every element of `L` has `TC` as its outer type constructor. 
+ */
+@implicitNotFound("Implicit not found: shapeless.UnaryTCConstraint[${L}, ${TC}]. Some element of ${L} does not have ${TC} as it's outer type constructor.")
+trait UnaryTCConstraint[L, TC[_]] extends Serializable
+
+trait LowPriorityUnaryTCConstraint1 {
+  private object UnaryTCConstraintDummy extends UnaryTCConstraint[Any, Id]
+
+  implicit def genericUnaryTC[H, HG, TC[_]](implicit
+    genH: Generic.Aux[H, HG],
+    utc: UnaryTCConstraint[HG, TC]) = UnaryTCConstraintDummy.asInstanceOf[UnaryTCConstraint[H, TC]]
+}
+
+trait LowPriorityUnaryTCConstraint0 extends LowPriorityUnaryTCConstraint1 {
+
+  implicit def hlistIdUnaryTC[L <: HList] = new UnaryTCConstraint[L, Id] {}
+  implicit def coproductIdUnarcyTC[C <: Coproduct] = new UnaryTCConstraint[C, Id]{}
+}
+
+trait LowPriorityUnaryTCConstraint extends LowPriorityUnaryTCConstraint0 {
```

Comment:
This trait seems unnecessary. Please merge into the companion object.

Reply:
Unfortunately, I can't get it to compile without the '0' and the other low priority.